File: Sources/CountriesFlagsObjC/source/Countries/UIImage+UnitedKingdom.m
//
//  UIImage+UnitedKingdom.m
//  CountriesFlags
//

#import "Countries/UIImage+UnitedKingdom.h"

@implementation UIImage (UnitedKingdom)

+ (UIImage *)flagUnitedKingdomWithSize:(CGSize)size {
    UIColor* colorBlue = [UIColor colorWithRed:1.0f/255.0f
                                         green:33.0f/255.0f
                                          blue:105.0f/255.0f
                                         alpha:1.0f];
    UIColor* colorRed = [UIColor colorWithRed:200.0f/255.0f
                                        green:16.0f/255.0f
                                         blue:46.0f/255.0f
                                        alpha:1.0f];
    UIColor* colorWhite = [UIColor whiteColor];
    
    UIGraphicsBeginImageContextWithOptions(size, YES, 0.0f);
    
    [colorBlue setFill];
    UIRectFill(CGRectMake(0.0f, 0.0f, size.width, size.height));
    
    [colorWhite setStroke];
    UIBezierPath* path = [UIBezierPath bezierPath];
    [path moveToPoint:CGPointZero];
    [path addLineToPoint:CGPointMake(size.width, size.height)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 6.0f;
    [path stroke];

    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height)];
    [path addLineToPoint:CGPointMake(size.width, 0.0f)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 6.0f;
    [path stroke];

    [colorRed setStroke];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 50.0f * -1.9f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 50.0f * 23.1f, size.height / 2.0f)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 2.0f;
    [path stroke];

    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 50.0f * 24.9f, size.height / 30.0f * 14.0f)];
    [path addLineToPoint:CGPointMake(size.width / 50.0f * 51.1f, size.height / 30.0f * -1.8f)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 2.0f;
    [path stroke];

    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 50.0f * 26.9f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(size.width / 50.0f * 51.9f, size.height)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 2.0f;
    [path stroke];

    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 50.0f * -1.9f, size.height / 30.0f * 32.3f)];
    [path addLineToPoint:(CGPointMake(size.width / 50.0f * 26.8f, size.height / 2.0f))];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 2.0f;
    [path stroke];

    [colorWhite setStroke];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 2.0f)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 10.0f;
    [path stroke];

    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 2.0f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0, size.height)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 10.0f;
    [path stroke];

    [colorRed setStroke];
    [path removeAllPoints];
    [path moveToPoint:CGPointMake(0.0f, size.height / 2.0f)];
    [path addLineToPoint:CGPointMake(size.width, size.height / 2.0)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 6.0f;
    [path stroke];

    [path removeAllPoints];
    [path moveToPoint:CGPointMake(size.width / 2.0f, 0.0f)];
    [path addLineToPoint:CGPointMake(size.width / 2.0f, size.height)];
    [path closePath];
    path.lineWidth = size.height / 30.0f * 6.0f;
    [path stroke];
    
    UIImage* image = UIGraphicsGetImageFromCurrentImageContext();
    UIGraphicsEndImageContext();
    
    return image != nil ? image : [UIImage new];
}

@end

File: Sources/CountriesFlags/Countries/UIImage+UnitedKingdom.swift
//
//  UIImage+UnitedKingdom.swift
//  CountriesFlags
//

#if canImport(UIKit)

import UIKit

extension UIImage {
    /**
     [wikipedia]: https://en.wikipedia.org/wiki/Union_Jack "wikipedia"
     
     Generates the flag of the United Kingdom.
     For more information see [wikipedia].

     - Parameters:
        - size: image size.
     
     - Returns: Returns the flag of the United Kingdom.
     */
    internal static func flagUnitedKingdom(with size: CGSize) -> UIImage? {
        let colorBlue = UIColor(red: 1.0 / 255.0,
                                green: 33.0 / 255.0,
                                blue: 105.0 / 255.0,
                                alpha: 1.0)
        let colorRed = UIColor(red: 200.0 / 255.0,
                               green: 16.0 / 255.0,
                               blue: 46.0 / 255.0,
                               alpha: 1.0)
        let colorWhite = UIColor.white

        UIGraphicsBeginImageContextWithOptions(size, false, 0.0)

        colorBlue.setFill()
        UIRectFill(CGRect(origin: .zero, size: size))

        colorWhite.setStroke()
        let path = UIBezierPath()
        path.move(to: .zero)
        path.addLine(to: CGPoint(x: size.width, y: size.height))
        path.close()
        path.lineWidth = size.height / 30.0 * 6.0
        path.stroke()

        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height))
        path.addLine(to: CGPoint(x: size.width, y: 0.0))
        path.close()
        path.lineWidth = size.height / 30.0 * 6.0
        path.stroke()

        colorRed.setStroke()
        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 50 * -1.9, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 50.0 * 23.1, y: size.height / 2.0))
        path.close()
        path.lineWidth = size.height / 30.0 * 2.0
        path.stroke()

        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 50 * 24.9, y: size.height / 30.0 * 14.0))
        path.addLine(to: CGPoint(x: size.width / 50.0 * 51.1, y: size.height / 30.0 * -1.8))
        path.close()
        path.lineWidth = size.height / 30.0 * 2.0
        path.stroke()

        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 50 * 26.9, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: size.width / 50.0 * 51.9, y: size.height))
        path.close()
        path.lineWidth = size.height / 30.0 * 2.0
        path.stroke()

        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 50 * -1.9, y: size.height / 30.0 * 32.3))
        path.addLine(to: CGPoint(x: size.width / 50.0 * 26.8, y: size.height / 2.0))
        path.close()
        path.lineWidth = size.height / 30.0 * 2.0
        path.stroke()

        colorWhite.setStroke()
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 2.0))
        path.close()
        path.lineWidth = size.height / 30.0 * 10.0
        path.stroke()

        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 2.0, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: size.height))
        path.close()
        path.lineWidth = size.height / 30.0 * 10.0
        path.stroke()

        colorRed.setStroke()
        path.removeAllPoints()
        path.move(to: CGPoint(x: 0.0, y: size.height / 2.0))
        path.addLine(to: CGPoint(x: size.width, y: size.height / 2.0))
        path.close()
        path.lineWidth = size.height / 30.0 * 6.0
        path.stroke()

        path.removeAllPoints()
        path.move(to: CGPoint(x: size.width / 2.0, y: 0.0))
        path.addLine(to: CGPoint(x: size.width / 2.0, y: size.height))
        path.close()
        path.lineWidth = size.height / 30.0 * 6.0
        path.stroke()

        let image = UIGraphicsGetImageFromCurrentImageContext()
        UIGraphicsEndImageContext()

        return image
    }
}

#endif
